torch/csrc/jit/import.cpp

```diff
@@ -319,4 +320,113 @@ std::shared_ptr<script::Module> ImportIRModule(
   return decoder.decode(module, serialized_module, storage_map);
 }
 
+// TODO: mmap file reader
+class PyTorchFileReader {
+ public:
+  PyTorchFileReader(const std::string& filename) {
+   fp = std::fopen(filename.c_str(), "rb");
+   std::fseek(fp, 0L, SEEK_END);
+   file_size = std::ftell(fp);
+   std::fseek(fp, 0L, SEEK_SET);
+   readAndValidateFileHeader();
+  }
+
+  // returns raw data map and the index of the last record (i.e. the ModelProto)
+  std::tuple<std::unordered_map<std::string, std::string>&, size_t> read_raw() {
+    JIT_ASSERT(!finalized);
+    size_t final_record_offset = cursor; // TODO: this is probably wrong
+    while (cursor != file_size) {
+      final_record_offset = readRecordIntoMap();
+    }
+    finalized = true;
+    return {raw_values, final_record_offset};
+  }
+
+  ~PyTorchFileReader() {
+    std::fclose(fp);
+  }
+
```

Comment:
`shared_ptr<void> readLastEntry()`

Reply:
I think I'm going to scrap the concept of assuming that the last entry is the proto and instead encode that as an offset in the file header